### ISSUE TITLE
Poisson NLL loss

### ISSUE THREAD

**Issue by vlasenkov** [pytorch/pytorch#1774]
There is no log-likelihood loss with underlying target's Poisson distribution. Would be great to add. I can do it.

**Comment by vlasenkov** [comment]
I'm not sure about where implementations of different losses are stored. To create new loss I need:

1.  to add new class to `torch/nn/modules/loss.py`
2. then register implementation of the loss somewhere in `torch/nn/_functions/thnn`
3. But what are the locations for these implementations? `torch/legacy` or `torch/nn/functional.py` or `torch/nn/_functions/loss.py` or some C code?

**Comment by fmassa** [comment]
The C/Cuda implementation can be found in `libs/THNN` and `libs/THCUNN`, which are automatically wrapped in python via `torch/nn/_functions/thnn/auto.py`. If the function can be trivially implemented using autograd, the implementation could go to `torch/nn/functional`, followed by a wrapper in the modules. If you want to write the backprop by hand and in python (for example for optimization reasons), then you could write it in `torch/nn/_functions/loss.py` for example.
